spec/memo_wise_spec.rb

```diff
@@ -3,1071 +3,917 @@
 require "values"
 
 RSpec.describe MemoWise do
-  let(:class_with_memo) do
-    Class.new do
-      prepend MemoWise
+  ##
+  # Shared contexts for setting up for testing:
+  #   * Instance methods
+  #   * Class methods defined via: 'def self.foo ...'
+  #   * Class methods defined via: 'class << self; def foo ...'
+  ##
 
-      def no_args_counter
-        @no_args_counter || 0
-      end
+  shared_context "with context for instance methods" do
+    let(:instance) { class_with_memo.new }
 
-      def with_positional_args_counter
-        @with_positional_args_counter || 0
-      end
+    let(:class_with_memo) do
+      Class.new do
+        prepend MemoWise
 
-      def with_positional_and_splat_args_counter
-        @with_positional_and_splat_args_counter || 0
-      end
+        DefineMethodsForTestingMemoWise.define_methods_for_testing_memo_wise(
+          target: self,
+          via: :instance
+        )
 
-      def with_keyword_args_counter
-        @with_keyword_args_counter || 0
-      end
+        def protected_memowise_method_counter
+          @protected_memowise_method_counter || 0
+        end
 
-      def with_keyword_and_double_splat_args_counter
-        @with_keyword_and_double_splat_args_counter || 0
-      end
+        def protected_memowise_method
+          @protected_memowise_method_counter =
+            protected_memowise_method_counter + 1
+          "protected_memowise_method"
+        end
+        protected :protected_memowise_method
+        memo_wise :protected_memowise_method
 
-      def with_positional_and_keyword_args_counter
-        @with_positional_and_keyword_args_counter || 0
-      end
+        def self.class_positional_args_counter
```

Comment:
Comments explaining why these methods are here would be really helpful.

Reply:
Comments added!